feat: Upstream Sync with Content Library Blocks

kdmccormick opened the pull request:
## Description

This is the majority of the backend work needed for the Content Libraries Beta launch in Sumac.

### The Python API

The Python API is entirely defined in cms/lib/xblock/upstream_sync.py, and features:

* A new UpstreamSyncMixin which is added to all CMS blocks, including:
  * `upstream` (e.g., the usage key of the library block)
  * `upstream_version` (e.g., the last-synced library block version number)
  * `upstream_display_name` (e.g., the library-defined title of the block)
  * `upstream_max_attempts` (e.g., the library-defined max-attempts for the block)
  * `get_customizable_fields(...)`, allowing blocks to set up further `upstream_*` fields.

* The `UpstreamLink` data object, which holds metadata on a block's relationship with its upstream, with classmethods for fetching the metadata.

* Three public API functions:
  * `sync_from_upstream` (e.g., update a block from its library block) 
  * `fetch_customizable_fields` (e.g., update the library-defined defaults, without updating visible content).
  * `decline_sync` (e.g., decline an update from a library block)

### The HTTP API

It is defined and documented here: https://github.com/kdmccormick/edx-platform/blob/kdmccormick/upstream-proto/cms/djangoapps/contentstore/rest_api/v2/views/downstreams.py#L2

### Temporary UI

This PR includes @bradenmacdonald 's basic, dev-only UI from this PR: https://github.com/kdmccormick/edx-platform/pull/16 . We will improve this UI before the Sumac launch, but having it landed will aid with testing and further development over the next week.

We display an icon when content comes from a content library (see on the left):

![Screenshot](https://github.com/user-attachments/assets/876f3dec-c5e9-4e31-ad8f-06bc50bf7c7a)

Or if the link is broken, it says that:

![Screenshot](https://github.com/user-attachments/assets/8ed17209-f424-4e22-9e38-55eff4496999)

And when an update is available, it displays a working "Update" button

![Screenshot](https://github.com/user-attachments/assets/a66b4b39-a02e-4199-9140-dbd2af095204)

It does not yet have a modal to allow you to accept/reject/preview the changes.

### Known issues

Definitely to be fixed before Sumac:

* Problems pasted from a library to a course are "converted" the advanced editor, even if they've only been edited with the visual editor.
* These UI, its icons, and copytext are temporary
* We sync from draft versions instead of published versions. See: #35626

Might be fixed before Sumac, definitely before Teak:

* List of customizable fields is super minimal right now (just display_name, max_attempts).
* API docs need to be moved from a docstring to decorators so they show up in the /api-docs/ UI. See #35653 

## Supporting Information

ADR: https://github.com/kdmccormick/edx-platform/blob/kdmccormick/upstream-proto/docs/decisions/0020-upstream-block.rst

## Testing 

### Basic setup (do this first)

* Create a (v2) content library with a component
* Copy the component from a library to a course unit
* Ensure: the little "Sourced from a library" icon shows up, but "Update Available" does NOT show up.

### Edit + Update

* Edit the library component.
* Refresh the course unit page.
  * Ensure: "Update Available" shows up
* Click "Update Available" to sync from the library
  * Ensure: Page reloads, with updated library block. "Update Available" button is gone.

### Overrides

* Edit the library component including updates to both:
  * a _customizable_ field, like the title or max_attempts
  * a _non-customizable_ field, like the problem/html data, or the video URL
* Refresh the course unit page
  * Ensure: "Update Available" shows up
* Edit the course-local instance of your library component, including an edit to both a customizable and a non-customizable field (ideally the same ones you edited above)
* Click "Update Available" to sync from the library
  * Ensure: Page reloads, with updated library block. "Update Available" button is gone.
  * Ensure: Your course-local edit to the customizable field remains intact.
  * Ensure: Your course-local edit to the non-customizable is blown away, and instead you see the library-level edit manifest

### "Restore to library default" (Sumac version)

In Sumac, we won't have full "Restore a field to its library default" functionality. We'll have a halfway version instead. [More details](https://github.com/kdmccormick/edx-platform/blob/c946d4d5514e9fa75f40cee2f914f1ba15a3d243/cms/lib/xblock/upstream_sync.py#L384). Here's how to test the halfway version, using display_name (although any customizable field should work) and a linked library/course component pair:

* In your library: Set your component's title to `My Upstream Title 1`
* In your course: Click "Update Available"
  * Ensure that the course-local title becomes `My Upstream Title 1`.
* In your course: Set your component's title to `My Downstream Title`
* In your library: Now set your component's title to `My Upstream Title 2`
* In your course: Click "Update Available"
  * Ensure that the course-local title **remains** `My Downstream Title`.
* In your course: Set your component's title to `My Upstream Title 2` _(this is our halfway "restore default" mechanism)_
* In your library: Now set your component's title to `My Upstream Title 3`
* In your course: Click "Update Available"
  * Ensure that the course-local title becomes `My Upstream Title 3`.

### Export/Import

* Edit the library component including updates to both:
  * a _customizable_ field, like the title or max_attempts
  * a _non-customizable_ field, like the problem/html data, or the video URL
* Refresh the course unit page
  * Ensure: "Update Available" shows up
* Export your course to OLX
* Import the OLX back into the same course run, and navigate to your library-linked component.
  * Ensure: "Update Available" is shown, and clicking it works as expected.
* Now import the OLX into a _different course run_ on the same instance, and navigate to your library-linked component in that course run.
  * Ensure: "Update Available" is shown, and clicking it works as expected.
* Now import the OLX into a _different instance_, and navigate to your library-linked component.
  * Ensure: The content says it is has a broken library link.

## Merge Deadline

ASAP

kdmccormick commented:
@bradenmacdonald So I've got this draft version of upstream sync working, implemented with XBlock fields and handlers. Using the XBlock fields and handlers for this was nice because it gave me persistence, URL routing, and OLX serialization for free.

But now I'm wondering if this is bad approach because we'll need to support Units and Subsections in libraries next, and we've agreed that those structural blocks should stop being XBlocks eventually.

The alternative I see is to:
* use Django models to store upstream links, upstream defaults, and upstream overrides,
* define REST/Python APIs for managing them, and
* handle OLX serialization as special case

Thoughts?

kdmccormick commented:
> use Django models to store upstream links, upstream defaults, and upstream overrides,

Okay, I started prototyping this, and it's easier said than done. Particularly, if you imagine a model like this, there's a question of whether we FK to entities or entityverisons:
```python
class UpstreamDownstreamLink(models.Model):
    """
    Link a piece of content (the downstream) to a source for updates (the upstream).
    """
    # Pretty straightforward -- FK to a particular upstream version, and then update the
    # version whenever sync happens.
    upstream = models.ForeignKey(PublishableEntityVersion)

    # Which one of these is right?? Neither?

    # This one would apply to all versions of the downstream component, so it would play weird
    # draft/publish, and would fail if we ever added an "undo" button to studio.
    downstream = models.OneToOneField(PublishableEntity)

    # This seems more correct, but would require updating every time a new version of the downstream
    # component is created. Perhaps this would work fine as a cache, but it seems wrong as the authoritative
    # data source.
    downstream = models.OneToOneField(PublishableEntityVersion)
```

It seems to me, then, that we *do* want to persist upstream metadata using XBlocks fields, so that version management works as we expect. Still, I'm open to feedback on whether we should use XBlock handlers vs. build a separate REST API for getting/settings links and syncing from upstream.

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7365186827.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7365186827-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7365186827-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7365186827-tutor-requirements.txt)

bradenmacdonald commented:
@kdmccormick I think it's fine to implement this with fields and handlers for now; as you said, it gives you a lot for free. It even gives us most of the Unit and Subsection functionality for free, I think. Yes, we want to implement those differently in the future but we aren't going to be able to make that change anytime soon, and I think we can worry about how library sync works in that case _after_ we have the new design for how that stuff works. Right now it's mostly an unknown.

e.g. I have an idea for an "outline" object that encodes the course hierarchy and library content and much more, and if we implemented it that way, we'd have to re-implement even your django model based prototype. Or previously I had proposed [this compositor architecture](https://openedx.atlassian.net/wiki/spaces/AC/pages/803405867/Compositor+Architecture+Proposal) which is different yet again. So I think it's not worth planning for that too much until we've got an ADR and buy-in about what the future direction is for the non-XBlock way to represent hierarchy.

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-34925-7daceb.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-34925-7daceb.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7439611288-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7439611288-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7439611288-tutor-requirements.txt)

kdmccormick committed:
feat: Upstream Sync with Content Library Blocks

This introdues the idea of "upstream" and "downstream" content,
where downstreams (like course components) can pull content updates from
upstreams (like learning core-backed content library blocks). This
supports the upcoming Content Libraries Relaunch Beta for Sumac.
New features include:

* A new XBlockMixin: UpstreamSyncMixin.
* A new CMS Python API: cms.lib.xblock.upstream_sync
* A new CMS JSON API: /api/contentstore/v2/downstreams
* A temporary, very basic UI for syncing from Content Library blocks

Implements:
https://github.com/kdmccormick/edx-platform/blob/kdmccormick/upstream-proto/docs/decisions/0020-upstream-block.rst

Co-Authored-By: Braden MacDonald <braden@opencraft.com>

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-34925-7daceb.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-34925-7daceb.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7441506907-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7441506907-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7441506907-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-34925-7daceb.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-34925-7daceb.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7453656289-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7453656289-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7453656289-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-34925-7daceb.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-34925-7daceb.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7464527066-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7464527066-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7464527066-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7475225430.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7475225430-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7475225430-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7475225430-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-34925-7daceb.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-34925-7daceb.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7488737979-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7488737979-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7488737979-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-34925-7daceb.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-34925-7daceb.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7495438114-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7495438114-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7495438114-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7496107434.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7496107434-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7496107434-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7496107434-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7517939999.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7517939999-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7517939999-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7517939999-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7518280560.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7518280560-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7518280560-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7518280560-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-34925-7daceb.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-34925-7daceb.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7521107834-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7521107834-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7521107834-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7558879893.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7558879893-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7558879893-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7558879893-tutor-requirements.txt)

ormsbee commented:
I think that it'll be important for LearningPackages to have independent lifecycles. Foreign keys across LearningPackage boundaries are fine, but they should have `on_delete=models.SET_NULL`, and behave properly when the originating LearningPackage is deleted for whatever reason.

So for instance, if we are storing "Problem C1 in my course is based on version 4 of Problem L1 in a library that lives in another LearningPackage," we need to store "version 4" in a data model that's associated with the course, and not rely on always being able to follow the foreign key to get the upstream version number.

> ```python
>     # Which one of these is right?? Neither?
> 
>     # This one would apply to all versions of the downstream component, so it would play weird
>     # draft/publish, and would fail if we ever added an "undo" button to studio.
>     downstream = models.OneToOneField(PublishableEntity)
> 
>     # This seems more correct, but would require updating every time a new version of the downstream
>     # component is created. Perhaps this would work fine as a cache, but it seems wrong as the authoritative
>     # data source.
>     downstream = models.OneToOneField(PublishableEntityVersion)
> ```

I think `downstream` is the `PublishableEntityVersion`. Some points in its favor:

* It will do the right thing when it comes to draft/publish and reverts, as you point out.
* It's plausible that a publishable entity will switch upstreams later in its life, e.g. if a library is removed and another one is re-added. Tracking each version separately makes that easier.
* We can model this like we would model other metadata that we would attach to a `PublishableEntityVersion`, in an established pattern that will be easier to cleanup.

To that last point, if we think of "upstream" as just another piece of extension-point metadata around the `PublishableEntityVersion`, then the model could look something like this:

```python
class VersionUpstream(models.Model):
    version = models.OneToOneField(PublishableEntityVersion, on_delete=models.CASCADE, primary_key=True)
    upstream_version = models.ForeignKey(PublishableEntityVersion, on_delete=models.SET_NULL)
    upstream_key = key_field()
    upstream_version_num = models.PositiveIntegerField()
```

Of course, all of that is in the scenario where downstream is in a Learning Core data model, which is not the case today. I think that for any model in the short term, we would:

* Have a similar model that just has a UsageKey for downstream (so no versioning).
* Store the source of truth for upstream links in the ModuleStore, and;
* Update the model whenever that ModuleStore entry changes.

So that model really only exists to answer the question, "What needs updates?"

kdmccormick commented:
Thanks @ormsbee . Agreed all around, particularly on:

> Store the source of truth for upstream links in the ModuleStore
> ...
> So that model really only exists to answer the question, "What needs updates?"

My latest commits get rid of the Django models. I figure we can add those as we get closer to needing them. What do you think (overall) of the current implementation?

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7591853246.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7591853246-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7591853246-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34925-7daceb-51564656-7591853246-tutor-requirements.txt)

ormsbee commented on cms/djangoapps/contentstore/helpers.py:
```diff
@@ -375,10 +385,20 @@ def _import_xml_node_to_parent(
+        try:
+            validate_upstream_key(copied_from_block)
+        except UnsupportedUpstreamKeyType:
+            pass  # @@TODO - should we let this error bubble up?
```
This would get called during course import, right? If so, maybe just log this as an error for now? At some point we may add other supported key types that people will have in their Teak-generated content exports. Sumac might not recognize those or be able to hook it up correctly when importing, but it should try not to explode the import process.

kdmccormick replied:
This is no longer necessary. The OLX has all the information the linked content needs.

ormsbee commented on cms/lib/xblock/upstream_sync.py:
```diff
@@ -0,0 +1,228 @@
+                continue
+            value = getattr(upstream_block, field_name)
+            if field.scope == Scope.settings:
+                self.upstream_settings[field_name] = value
```
I think we should do that thing we talked about where we have a fixed set of known overridable fields, and not generally pull from the settings scope. Otherwise we're going to make pretty much the entire VideoBlock overridable:

https://github.com/openedx/edx-platform/blob/ca46c20abb646b8b887e3c4d999f4e069f7eb871/xmodule/video_block/video_xfields.py#L33-L122

And we're going to run into that weird issue where the ProblemBlock's `markdown` field is settings scope, but changes to it try to write to the content scoped `data` field and just mess everything up:

https://github.com/openedx/edx-platform/blob/ca46c20abb646b8b887e3c4d999f4e069f7eb871/xmodule/capa_block.py#L323

kdmccormick replied:
Done (`get_customizable_fields`)

kdmccormick commented on cms/lib/xblock/upstream_sync.py:
```diff
@@ -0,0 +1,228 @@
+        for field_name, value in self.upstream_settings.items():
+            if field_name not in self.upstream_overridden:
+                if value != getattr(self, field_name):
+                    self.upstream_overridden.append(field_name)
```
```suggestion
                    self.upstream_overridden.add(field_name)
```
upstream_overridden is now a Set, not a List

kdmccormick replied:
Punted

ormsbee commented:
> What do you think (overall) of the current implementation?

Overall, I think the approach is good–especially given where we are today. My only review comments are on lower-level details.

bradenmacdonald commented on cms/djangoapps/contentstore/helpers.py:
```diff
@@ -375,10 +385,21 @@ def _import_xml_node_to_parent(
+        # If requested, link this block as a downstream of where it was copied from
+        temp_xblock.upstream = copied_from_block
+        try:
+            sync_from_upstream(temp_xblock, user, apply_updates=False)
```
I would appreciate a comment above this `sync_from_upstream` line that says what it's supposed to do here. Because it's not actually syncing, right? It's (I think) mostly setting the important `upstream_version` field, but also storing the original values of the upstream fields ?

I think we also need to call out a common case that needs to be handled carefully: if the pasted block is from a draft version of the block, we have actually pasted a newer version than is currently available upstream. In that case, we don't want `downstream.upstream_version = upstream_link.version_available` but rather `downstream.upstream_version = copied_from_block.draft_version` or something like that. (I know we won't necessarily support that in this PR before merging but we should still call it out here in a comment).

kdmccormick replied:
@bradenmacdonald Thanks for calling this out. I was trying to treat this as a special case of `sync_from_upstream`, but that's really not right--it's not a sync at all. All we want to do after pasting is:
* Set the upstream to the copied block's usage key
* Set the upstream_version to the copied block's version. I added a `version_num` field to `StagedContentData` to handle this (it'll be 0 for things without versions, like course blocks).
* Fetch the upstream-set values for customizable fields, and save them into the pasted block's `upstream_*`fields. I've factored out a new function for this, although I'm thrashing on the name a bit. I think I'll call it `fetch_upstream_values_for_customizable_fields(...)` unless you have any objections.

bradenmacdonald replied:
Sounds good!

bradenmacdonald commented on cms/djangoapps/contentstore/xblock_storage_handlers/view_handlers.py:
```diff
@@ -539,7 +539,9 @@ def _create_block(request):
-                parent_key=usage_key, request=request
+                parent_key=usage_key,
+                request=request,
+                link_to_upstream=request.json.get("link_to_upstream"),
```
I think what we discussed recently was that instead of prompting the user to ask if they want to link or not, we should just always assume `link_to_upstream=True` if pasting from a library, and authors can just break the link if they don't want it linked.

kdmccormick replied:
Yup, you're right. Done.

rpenido commented:
Hi @kdmccormick! :smiley: 

I'm working on the frontend to implement this:
- https://github.com/openedx/frontend-app-authoring/issues/1337

I'm a bit lost with the naming here, so I would like to confirm:
When we add a library component to a course, we first need to copy the component to the course (creating a "downstream" block) and then link it to the block from the library ("upstream¨), right?

Do we already have a REST endpoint that does this?

I only found the `PUT /api/v2/contentstore/downstreams/{usage_key_string}`, which seems to do just the link step.

bradenmacdonald commented:
> I'm a bit lost with the naming here, so I would like to confirm:
> When we add a library component to a course, we first need to copy the component to the course (creating a "downstream" block) and then link it to the block from the library ("upstream¨), right?

@rpenido That's right. The source block (in the library) is the upstream, and the linked copy (in the course) is the "downstream". For testing purposes, if you have this branch checked out, copying a block from a library and pasting it into a course should do the same thing (create the linked downstream copy) that we want to do via the other "Add library content to a course" workflow that you're working on. Although last time I tried this branch, it wasn't working due to a missing migration.

ormsbee commented on cms/djangoapps/contentstore/helpers.py:
```diff
@@ -375,10 +384,30 @@ def _import_xml_node_to_parent(
+            # this could be the latest Published version, or it could be an an even newer Draft version.
+            temp_xblock.upstream_version = copied_from_version_num
+            # Also, fetch upstream values (`upstream_display_name`, etc.).
+            # Recall that the copied block could be a draft. So, rather than fetching form the published upstream (which
```
```suggestion
            # Recall that the copied block could be a draft. So, rather than fetching from the published upstream (which
```

ormsbee commented on cms/djangoapps/contentstore/rest_api/v2/views/downstreams.py:
```diff
@@ -0,0 +1,228 @@
+        """
+        Inspect an XBlock's link to upstream content.
+        """
+        assert isinstance(request.user, AbstractUser)
```
[question]: What is this assert for?

kdmccormick replied:
These asserts were to convince mypy that request.user is not an AnonymousUser. I've found a nicer way to do it without any asserts, in my latest commit.

ormsbee commented on cms/lib/xblock/upstream_sync.py:
```diff
@@ -0,0 +1,395 @@
+    For each customizable field `downstream.blah`:
+
+    1. Fetch `upstream.blah` and save it in `downstream.upstream_blah`.
+    2. If not only_fetch, and if `downstream.blah` isn't customized, update it from `upstream.blah`.
```
[Nit]: Please change this to a realistic, concrete example with real field names, framing a block in a library as upstream and a block in a course as downstream.

kdmccormick replied:
Will do in a follow-up

kdmccormick replied:
Actually, done

ormsbee commented on cms/djangoapps/contentstore/rest_api/v2/views/downstreams.py:
```diff
@@ -0,0 +1,227 @@
+            if sync_param == "true":
+                sync_from_upstream(downstream=downstream, user=request.user)
+            else:
+                fetch_customizable_fields(downstream=downstream, user=request.user)
```
[Nit]: Calling `sync_from_upstream` when `sync_param` is true is intuitive, but you might want to put a small explanation for why `fetch_customizable_fields` is called when `sync_param` is 'false'.

ormsbee replied:
<strike>[question]: is this where we update the version number that we're syncing from?</strike>

kdmccormick replied:
> [Nit]: Calling sync_from_upstream when sync_param is true is intuitive, but you might want to put a small explanation for why fetch_customizable_fields is called when sync_param is 'false'.

Will do in a follow-up 👍🏻 

> [question]: is this where we update the version number that we're syncing from?

Sorry, can you elaborate?

ormsbee replied:
You can ignore that last one–I realized how version bumping happened later on in the PR, but forgot to remove this question.

kdmccormick replied:
Nit is resolved too

ormsbee commented on cms/djangoapps/contentstore/rest_api/v2/views/downstreams.py:
```diff
@@ -0,0 +1,227 @@
+            raise ValidationError({"upstream_ref": str(exc)}) from exc
+        modulestore().update_item(downstream, request.user.id)
+        link = UpstreamLink.get_for_block(downstream)
+        assert link
```
Do you plan to keep this assertion?

kdmccormick replied:
This was also to make mypy happy-- it wanted us to check for None since `get_for_block` will return None iff `downstream.upstream is None`, but we've ascertained above that `downstream.upstream` is non-None, so it's a spurious type warning.

I changed it in my latest commit to just use a `# type: ignore[assignment]` rather than an assert.

kdmccormick replied:
Based on [these changes](https://github.com/openedx/edx-platform/pull/34925#discussion_r1801586472), I was able to make `get_for_block` return `UpstreamLink` (non-nullable), which made this type-ignore stuff unneccessary.

ormsbee commented on cms/djangoapps/contentstore/rest_api/v2/views/downstreams.py:
```diff
@@ -0,0 +1,227 @@
+    def put(self, request: Request, usage_key_string: str) -> Response:
+        """
+        Edit an XBlock's link to upstream content.
+        """
```
Please describe the use case for why a client would call this in the docstring.

kdmccormick replied:
Done

ormsbee commented on cms/djangoapps/contentstore/rest_api/v2/views/downstreams.py:
```diff
@@ -0,0 +1,227 @@
+        assert isinstance(request.user, AbstractUser)
+        downstream = _load_accessible_block(request.user, usage_key_string, require_write_access=True)
+        new_upstream_ref = request.data.get("upstream_ref")
+        if not isinstance(new_upstream_ref, str):
```
[edge-case question] Is there ever a good case for changing the upstream reference to be some null value, i.e. differentiating between "this is no longer a syncable thing" from "this is a syncable thing, but it's not syncing to the _right_ thing"? Or is that covered by just deleting the downstream and making a new entry later?

kdmccormick replied:
I'm not sure I follow. Do you have a more concrete example in mind?

ormsbee replied:
I think this is probably my lack of understanding.

Question: In terms of the API and data model, what is the difference between:

1. This course component used to sync to a library, but I want to break that link so it never asks me to sync again, and it actually doesn't even think this was from a library any longer. (Maybe because I never meant for it to be something that syncs.)
2. This course component used to sync to a library, but that library is not available on this site at this time.

kdmccormick replied:
Gotcha. In the data model:

1. An un-linked block has `block.upstream is None`. So, to sever a block's link, we just set `block.upstream = None`. Calling any of the API methods (`UpstreamLink.get_for_block`, `sync_from_upstream`, `fetch_customizable_fields`) would then return None / be a no-op.
2. If the library is unavailable, `block.upstream` would remain set to the library's key, BUT calling any of the API methods (`UpstreamLink.get_for_block`, `sync_from_upstream`, `fetch_customizable_fields`) will raise a BadUpstream. Calling `UpstreamLink.try_get_for_block` will yield an `UpstreamLink` with a non-null `error_message`.

Does that make sense?

Now, in the HTTP API, I'm realizing that the return codes for each of these scenarios is a bit inconsistent from endpoint to endpoint:

1. unlinked blocks return 400
2. bad links sometimes return 200 with an `error_message`, sometimes return 400

I'm going to think a bit on the HTTP API response codes and push a revision to it tomorrow.

kdmccormick replied:
Okay @ormsbee , I have pushed a new commit, which changes things a bit.

1. Blocks with no link (or a severed link):
   * In the data model, this is still represented as `block.upstream is None`
     * If the block was previously linked to an upstream, that will be reflected on`block.copied_from_block`.
   * calling any upstream_sync APIs raises NoUpstream
   * calling `GET /api/contentstore/v2/downstreams/{block.usage_key}` yields a 200 with an `error_message` and log entries explaining that the upstream link is missing.
   * calling `POST /api/contentstore/v2/downstreams/{block.usage_key}/sync` yields a 400 with an `error_message` and log entries explaining that the upstream link is missing.

2. Blocks with an malformed link and/or a link to an unavailable upstream:
   * In the data model, `block.upstream` will be set.
   * calling any upstream_sync APIs raises BadUpstream.
   * calling `GET /api/contentstore/v2/downstreams/{block.usage_key}` yields a 200 with an `error_message` and log entries explaining that the upstream link is invalid.
   * calling `POST /api/contentstore/v2/downstreams/{block.usage_key}/sync` yields a 400 with an `error_message` and log entries explaining that the upstream link is missing.

I feel much better about this API, and it ended being net fewer lines of code. Let me know what you think.

ormsbee replied:
I think this is great, btw.

ormsbee commented on cms/lib/xblock/upstream_sync.py:
```diff
@@ -0,0 +1,395 @@
+    """
+    Allows an XBlock in the CMS to be associated & synced with an upstream.
+
+    Mixed into CMS's XBLOCK_MIXINS, but not LMS's.
```
[nit/out-of-scope discussion]: I think we'll eventually want to emit a block's upstream information in tracking log events, and thus need to expose some of this to the LMS. Definitely not in scope here though.

kdmccormick replied:
Yeah, I'm thinking about that now over in this PR: https://github.com/openedx/edx-platform/pull/35553

I might end up mixing it into the LMS too, but I'll leave this here for now.

ormsbee commented on cms/lib/xblock/upstream_sync.py:
```diff
@@ -0,0 +1,395 @@
+    # but we need to do more due diligence to be sure that they cover all cases, including API edits, import/export,
+    # copy/paste, etc. We will figure this out in time for the full Content Libraries Relaunch (related ticket:
+    # https://github.com/openedx/frontend-app-authoring/issues/1317). But, for the Beta realease, we're going to
+    # implement something simpler:
```
[off-topic, out-of-scope]: 😞 Do we have to hook in all the way down at the Field Data layer?

kdmccormick replied:
I'm not sure yet. It might be good to straighten out `XBlock.editor_saved` and make it a proper XBlock API which we can rely upon here.

ormsbee commented on cms/envs/common.py:
```diff
@@ -1008,6 +1008,7 @@
     XModuleMixin,
     EditInfoMixin,
     AuthoringMixin,
+    'cms.lib.xblock.upstream_sync.UpstreamSyncMixin',
```
Please explain in comments why this is specified as a string instead of being imported like the rest. (Circular dependency issue?)

kdmccormick replied:
Kinda. It was that we can't load any djangoapps from settings files, and `lib/upstream_sync.py` imported from a couple djangoapps.

In my latest commit, I went back and made this a proper import, and then changed move some upstream_sync.py imports into their functions to avoid the problem, each with an explanatory comment.

ormsbee commented on cms/templates/studio_xblock_wrapper.html:
```diff
@@ -103,7 +108,22 @@
-                <span class="xblock-display-name">${label}</span>
+                    % if upstream_info:
+                        % if upstream_info.error_message:
+                            <!-- "broken link" icon from https://fonts.google.com/icons?selected=Material+Symbols+Outlined:link_off:FILL@0;wght@400;GRAD@0;opsz@24&icon.size=24&icon.color=%235f6368&icon.query=link -->
```
Is this because there's no Paragon equivalent?

bradenmacdonald replied:
Actually there is, the LinkOff icon in Paragon, but I don't think we have a way to display Paragon icons in this legacy UI, and they won't match any of the other (Font Awesome) icons anyways. When we implement this "properly" in the MFE, we'll use a Paragon icon.

ormsbee commented on cms/templates/studio_xblock_wrapper.html:
```diff
@@ -103,7 +108,22 @@
+                            <span class="sr-only">${_("Sourced from a library - but the upstream link is broken/invalid.")}</span>
+                        % else:
+                            <!-- "library" icon from https://fonts.google.com/icons?selected=Material+Symbols+Outlined:newsstand:FILL@0;wght@400;GRAD@0;opsz@24&icon.size=24 -->
+                            <svg role="img" data-tooltip="${_("Sourced from a library.")}" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" width="24px" height="24px" viewBox="0 -960 960 960" fill="currentColor" style="vertical-align: middle; padding-bottom: 4px;">
```
Likewise, is there no Paragon equivalent?

bradenmacdonald replied:
This is the icon I was told to use and there is no Paragon equivalent. It's easy enough to use this in the legacy UI, and the legacy UI is not using Paragon icons anyways (it uses Font Awesome).

When we implement this in the MFE Unit Page, we will either add this icon to Paragon or choose a Paragon icon instead.

Closest paragon icons:
![Screenshot](https://github.com/user-attachments/assets/08b9fb6d-5dbf-41ee-a627-21637ca33b1e)

ormsbee commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -243,6 +246,8 @@ def from_component(cls, library_key, component):
             display_name=component.versioning.draft.title,
             created=component.created,
             modified=component.versioning.draft.created,
+            draft_version_num=component.versioning.draft.version_num,
```
We already have a reference to `draft` above, so I think this can be `draft_version_num=draft.version_num`. Likewise with the `modified` param (which I realize isn't part of this PR).

kdmccormick replied:
Fixed in the latest commit

ormsbee commented on cms/templates/studio_xblock_wrapper.html:
```diff
@@ -114,6 +134,18 @@
+                                <button
+                                    class="btn-default library-sync-button action-button"
+                                    data-tooltip="${_("Update available - click to sync")}"
+                                    onclick="$.post('/api/contentstore/v2/downstreams/${xblock.usage_key}/sync').done(() => { location.reload(); })"
```
[nit]: this seems fragile

bradenmacdonald replied:
This is my hacky UI code just to get something working. In the coming weeks I'll be replacing this with a proper modal dialog that lets you compare the new version to the previous version and decide whether or not you want to accept the changes. The backend part of that is already mostly working - https://github.com/openedx/edx-platform/pull/35626 . The frontend spec is at https://github.com/openedx/frontend-app-authoring/issues/1341 but I haven't started it yet.

ormsbee commented on cms/djangoapps/contentstore/rest_api/v2/views/downstreams.py:
```diff
@@ -0,0 +1,227 @@
+
+    def delete(self, request: Request, usage_key_string: str) -> Response:
+        """
+        Decline the latest updates to the block at {usage_key_string}.
```
[question]: If I call delete on this opportunity to sync, is there any way to come back to it later if I change my mind? Or do I need to wait until the upstream version bumps again?

kdmccormick replied:
Currently neither the API nor the UI has any way to "un-decline" a sync.

Theoretically, one could call `POST .../sync` in order to accept a previously-declined update.

ormsbee commented on cms/lib/xblock/upstream_sync.py:
```diff
@@ -0,0 +1,395 @@
+            # Will be fixed in a follow-up task: https://github.com/openedx/edx-platform/issues/35582
+            # version_available=(lib_meta.published_version_num if lib_meta else None),
+            version_declined=downstream.upstream_version_declined,
+            error_message=None,
```
[request]: The sync process in general seems like a place where things could easily start going wrong, and I'm a little concerned that these sorts of diagnostic messages will get swallowed up by app code without getting logged somewhere, making it difficult for to detect when things start breaking on sites. Please add some logging for the messages that indicate where things have gone unexpectedly wrong (as opposed to "courses are never upstreams" type of logic).

kdmccormick replied:
Good call. In my latest commit, `try_get_for_block` now calls `logger.exception` with some extra details.

kdmccormick replied:
Done

ormsbee commented on cms/djangoapps/contentstore/rest_api/v2/views/downstreams.py:
```diff
@@ -0,0 +1,227 @@
+    if require_write_access and not has_studio_write_access(user, usage_key.context_key):
+        raise not_found
+    if not has_studio_read_access(user, usage_key.context_key):
+        raise not_found
```
Should this log an error as well?

kdmccormick replied:
@ormsbee AFAICT, we don't typically do additional logging for cut-and-dry 403/404 scenarios, so I'm not sure why we would here.

ormsbee replied:
Fair point. The main reason I bring it up is because in this case, it might be confusing to see a bunch of 404s in the logs for this, when it's really an underlying authz check that's failing.

bradenmacdonald replied:
(Not asking for changes now.) Personally, I've really come to appreciate a clear separation between 403 and 404 errors for things like this. Makes debugging, messaging to users, and supporting users much easier. "Leaking" what IDs exist or do not exist in a course is not really a big security concern, and would only apply to users who somehow have permission to read some parts of the course but not other parts - a tiny subset of user requests.

kdmccormick replied:
@bradenmacdonald I do like to use the full range of expressiveness that HTTP gives us, for all the reasons you highlight. So I'm totally open to changing this to 403 if we deem it not to be an integrity issue. I'm leaving it for now, but I'll add it to my list of things to follow up on.

ormsbee reviewed (changes requested):
This looks really good. 😁 My review is mostly just questions and nits, with a handful of requests to expand on the comments.

The only high level thing I'd ask is that you please go back over it and see if there are places where there should be logging added so that we might be able to better diagnose errors in a production system (with whatever extra data would be needed for that, like logging User IDs, course keys, library keys, usage keys, etc.).

Thank you!

kdmccormick commented on cms/djangoapps/contentstore/helpers.py:
```diff
@@ -375,10 +384,30 @@ def _import_xml_node_to_parent(
         # Store a reference to where this block was copied from, in the 'copied_from_block' field (AuthoringMixin)
         temp_xblock.copied_from_block = copied_from_block
+        # Try to link the pasted block (downstream) to the copied block (upstream).
+        temp_xblock.upstream = copied_from_block
```
@bradenmacdonald Currently, this PR sets `.upstream` when applicable _in addition_ to `.copied_from_block`. Should it set `.upstream` _instead of_ `.copied_from_block`? It would make the OLX a little less verbose.

```xml
<html filename="html1" copied_from_block="lb:Kyle:Lib1:html:fbbc03fa-e8b0-4d71-88bf-87fead5f850b" display_name="Text v5" editor="visual" name="null" source_code="null" tags="[]" upstream="lb:Kyle:Lib1:html:fbbc03fa-e8b0-4d71-88bf-87fead5f850b" upstream_display_name="Text v5" upstream_version="5" use_latex_compiler="false"/>
```

bradenmacdonald replied:
@kdmccormick I think that makes sense, assuming that if/when we unlink it from the library, we remove `.upstream` and instead set `.copied_from_block` at that time.

kdmccormick replied:
Great idea. Done, in my latest commit.

kdmccormick commented:
Thanks @ormsbee . I'll address all your concerns Monday, but I'm going to leave the nits for a followup PR.

@bradenmacdonald this is ready for review

kdmccormick commented:
@ormsbee Ready for another pass

kdmccormick commented on cms/lib/xblock/upstream_sync.py:
```diff
@@ -274,17 +293,38 @@ def decline_sync(downstream: XBlock) -> None:
     downstream.upstream_version_declined = upstream_link.version_available
 
 
+def sever_upstream_link(downstream: XBlock) -> None:
+    """
+    Given an XBlock that is linked to upstream content, disconnect the link, such that authors are never again prompted
+    to sync upstream updates. Erase all `.upstream*` fields from the downtream block.
+
+    However, before nulling out the `.upstream` field, we copy its value over to `.copied_from_block`. This makes sense,
+    because once a downstream block has been de-linked from source (e.g., a Content Library block), it is no different
+    than if the block had just been copy-pasted in the first place.
+
+    Does not save `downstream` to the store. That is left up to the caller.
+
+    If `downstream` lacks a link, then this raises NoUpstream (though it is reasonable for callers to handle such
+    exception and ignore it, as the end result is the same: `downstream.upstream is None`).
+    """
+    if not downstream.upstream:
+        raise NoUpstream()
+    downstream.copied_from_block = downstream.upstream
+    downstream.upstream = None
+    downstream.upstream_version = None
+    for _, fetched_upstream_field in downstream.get_customizable_fields().items():
+        setattr(downstream, fetched_upstream_field, None)  # Null out upstream_display_name, et al.
+
+
```
@ormsbee Note that I've gone an implemented `sever_upstream_link` (and its matching HTTP API, `DELETE /api/contentstore/v2/downstream/{key}`. It's not needed for the MVP, but it was easy enough to do, and it ensures that we don't forget [Braden's copied_from_block suggestion here](https://github.com/openedx/edx-platform/pull/34925#discussion_r1797412093).

ormsbee commented on cms/djangoapps/contentstore/rest_api/v2/views/downstreams.py:
```diff
@@ -0,0 +1,251 @@
+API paths (We will move these into proper api_doc_tools annotations soon
+https://github.com/openedx/edx-platform/issues/35653):
+
+  /api/v2/contentstore/downstreams/{usage_key_string}
```
Should this be `/api/contentstore/v2/downstreams/{usage_key_string}` here (and elsewhere in this docstring)?

kdmccormick replied:
@ormsbee yup, will fix

kdmccormick replied:
Done

ormsbee reviewed (changes requested):
Just a minor question about the API endpoint in the docstring (whether it's intended to be `/api/contentstore/v2/...`). Please consider the remaining logging request as optional.

kdmccormick commented on cms/lib/xblock/upstream_sync.py:
```diff
@@ -0,0 +1,442 @@
+    # Try to load the upstream.
+    upstream_link = UpstreamLink.get_for_block(downstream)  # Can raise UpstreamLinkException
+    if not upstream_link:
+        return  # No upstream -> nothing to sync.
```
```suggestion
    upstream_link = UpstreamLink.get_for_block(downstream)  # Can raise UpstreamLinkException
```

Forgot to simplify this block, now that `get_for_block` is guaranteed to return non-None.

kdmccormick replied:
Done

bradenmacdonald commented on cms/djangoapps/contentstore/rest_api/v2/urls.py:
```diff
@@ -1,15 +1,31 @@
+        home.HomePageCoursesViewV2.as_view(),
         name="courses",
     ),
+    # TODO: Potential future path.
```
At some point, we'd like to be able to give an _upstream_ ID and get the list of all its downstreams. I'm assuming we'll need some kind of new data model to track that before we're able to implement something like that in the future, right?

kdmccormick replied:
Yep. Indeed, we'll need something like a `DownstreamUpstreamLink` table, which the upstream_sync API writes to and keeps updated.

bradenmacdonald commented on cms/lib/xblock/upstream_sync.py:
```diff
@@ -0,0 +1,451 @@
+            version_synced=downstream.upstream_version,
+            version_available=(lib_meta.draft_version_num if lib_meta else None),
+            # TODO: Previous line is wrong. It should use the published version instead, but the
+            # LearningCoreXBlockRuntime APIs do not yet support published content yet.
```
They do now :) (mostly)

kdmccormick replied:
Nice! One of us can do a follow-up right after this merges, then.

bradenmacdonald commented on cms/lib/xblock/upstream_sync.py:
```diff
@@ -0,0 +1,451 @@
+    If `downstream` lacks a valid+supported upstream link, this raises an UpstreamLinkException.
+    """
+    link = UpstreamLink.get_for_block(downstream)  # can raise UpstreamLinkException
+    # We import load_block here b/c UpstreamSyncMixin is used by cms/envs, which loads before the djangoapps are ready.
+    from openedx.core.djangoapps.xblock.api import load_block  # pylint: disable=wrong-import-order
```
It could be nicer in the future to move `UpstreamSyncMixin` to its own file? It seems fairly separable from everything else in this file.

kdmccormick replied:
Yeah, I'm also thinking that might be necessary if we're to make these fields available in the LMS (for analytics) without exposing the whole upstream_sync API to LMS.

I'll leave it for now, but I will likely do this refactoring when it becomes necessary.

bradenmacdonald reviewed (approved):
Awesome work, @kdmccormick and very nice clean code 💯

I didn't test import-export but I tested various other things and it was working as advertised: 
![Screenshot](https://github.com/user-attachments/assets/07521561-1a1c-4459-8683-d18829d0aab8)
